File: backend/src/modules/growth/growth.service.ts
import { Injectable, NotFoundException, BadRequestException } from '@nestjs/common';

export interface ReferralCode {
  id: string;
  userId: string;
  code: string;
  maxUses: number;
  currentUses: number;
  createdAt: Date;
  expiresAt: Date;
}

export interface Referral {
  id: string;
  referrerId: string;
  referrerName: string;
  inviteeId: string;
  inviteeName: string;
  referralCode: string;
  status: string;
  createdAt: Date;
  acceptedAt?: Date;
}

export interface UserLevel {
  level: number;
  title: string;
  minXp: number;
  maxXp: number;
  reward: { gems: number; frameId?: string; badgeId?: string };
}

export interface SeasonalEvent {
  id: string;
  title: string;
  description: string;
  type: string;
  status: string;
  startDate: Date;
  endDate: Date;
  goals: EventGoal[];
  reward: EventReward;
  participantsCount: number;
}

export interface EventGoal {
  id: string;
  title: string;
  targetValue: number;
  currentValue: number;
  unit: string;
}

export interface EventReward {
  xp: number;
  gems: number;
  badgeId?: string;
  frameId?: string;
}

@Injectable()
export class GrowthService {
  private referralCodes: Map<string, ReferralCode> = new Map();
  private referrals: Map<string, Referral> = new Map();
  private userXp: Map<string, number> = new Map();
  private events: Map<string, SeasonalEvent> = new Map();

  constructor() {
    this.initializeDemoData();
  }

  private initializeDemoData() {
    // Demo referral code
    const demoCode: ReferralCode = {
      id: 'ref_code_1',
      userId: 'user_1',
      code: 'LEXI2024',
      maxUses: 10,
      currentUses: 3,
      createdAt: new Date(),
      expiresAt: new Date(Date.now() + 30 * 86400000),
    };
    this.referralCodes.set(demoCode.code, demoCode);

    // Demo seasonal event
    const demoEvent: SeasonalEvent = {
      id: 'event_1',
      title: 'German Summer Challenge',
      description: 'Master German this summer with daily practice!',
      type: 'challenge',
      status: 'active',
      startDate: new Date(Date.now() - 5 * 86400000),
      endDate: new Date(Date.now() + 25 * 86400000),
      goals: [
        { id: 'goal_1', title: 'Speaking Practice', targetValue: 100, currentValue: 35, unit: 'minutes' },
        { id: 'goal_2', title: 'Vocabulary Building', targetValue: 500, currentValue: 180, unit: 'words' },
        { id: 'goal_3', title: 'Goethe Tasks', targetValue: 20, currentValue: 8, unit: 'exercises' },
      ],
      reward: { xp: 5000, gems: 1000, badgeId: 'summer_master', frameId: 'summer_frame' },
      participantsCount: 8500,
    };
    this.events.set(demoEvent.id, demoEvent);
  }

  // Referral System
  generateReferralCode(userId: string): ReferralCode {
    const code = `LEXI${userId.substring(0, 4).toUpperCase()}${Math.floor(Math.random() * 9999).toString().padStart(4, '0')}`;

    const referralCode: ReferralCode = {
      id: `ref_${Date.now()}`,
      userId,
      code,
      maxUses: 10,
      currentUses: 0,
      createdAt: new Date(),
      expiresAt: new Date(Date.now() + 30 * 86400000),
    };

    this.referralCodes.set(code, referralCode);
    return referralCode;
  }

  applyReferralCode(code: string, inviteeId: string, inviteeName: string): Referral {
    const referralCode = this.referralCodes.get(code);
    if (!referralCode) throw new NotFoundException('Invalid referral code');
    if (referralCode.currentUses >= referralCode.maxUses) throw new BadRequestException('Code usage limit reached');
    if (new Date() > referralCode.expiresAt) throw new BadRequestException('Code expired');

    referralCode.currentUses++;

    const referral: Referral = {
      id: `ref_${Date.now()}`,
      referrerId: referralCode.userId,
      referrerName: 'User',
      inviteeId,
      inviteeName,
      referralCode: code,
      status: 'accepted',
      createdAt: new Date(),
      acceptedAt: new Date(),
    };

    this.referrals.set(referral.id, referral);
    return referral;
  }

  getReferralStats(userId: string) {
    const userReferrals = Array.from(this.referrals.values()).filter(r => r.referrerId === userId);
    return {
      totalReferrals: userReferrals.length,
      successfulReferrals: userReferrals.filter(r => r.status === 'accepted').length,
      totalGemsEarned: userReferrals.filter(r => r.status === 'accepted').length * 500,
      totalPremiumDaysEarned: userReferrals.filter(r => r.status === 'accepted').length * 7,
    };
  }

  // Gamification
  getUserXp(userId: string): number {
    return this.userXp.get(userId) || 0;
  }

  addXp(userId: string, xp: number): { totalXp: number; level: number } {
    const currentXp = this.getUserXp(userId);
    const totalXp = currentXp + xp;
    this.userXp.set(userId, totalXp);

    return {
      totalXp,
      level: this.calculateLevel(totalXp),
    };
  }

  calculateLevel(xp: number): number {
    let level = 1;
    let requiredXp = 100;

    while (xp >= requiredXp) {
      level++;
      requiredXp += level * 50;
    }

    return level;
  }

  getLevels(): UserLevel[] {
    return [
      { level: 1, title: 'Beginner', minXp: 0, maxXp: 100, reward: { gems: 0 } },
      { level: 5, title: 'Explorer', minXp: 500, maxXp: 1000, reward: { gems: 100, frameId: 'explorer_frame' } },
      { level: 10, title: 'Adventurer', minXp: 2000, maxXp: 3000, reward: { gems: 200, frameId: 'adventurer_frame' } },
      { level: 15, title: 'Scholar', minXp: 5000, maxXp: 7500, reward: { gems: 300, badgeId: 'scholar_badge' } },
      { level: 20, title: 'Expert', minXp: 10000, maxXp: 15000, reward: { gems: 500, frameId: 'expert_frame' } },
      { level: 25, title: 'German Student', minXp: 20000, maxXp: 30000, reward: { gems: 750, badgeId: 'german_student_badge' } },
      { level: 30, title: 'Linguist', minXp: 40000, maxXp: 60000, reward: { gems: 1000, frameId: 'linguist_frame' } },
      { level: 40, title: 'Master', minXp: 100000, maxXp: 150000, reward: { gems: 2000, badgeId: 'master_badge' } },
      { level: 50, title: 'Language Master', minXp: 250000, maxXp: 300000, reward: { gems: 5000, frameId: 'master_frame' } },
    ];
  }

  // Seasonal Events
  getEvents(): SeasonalEvent[] {
    return Array.from(this.events.values());
  }

  joinEvent(eventId: string, userId: string): SeasonalEvent {
    const event = this.events.get(eventId);
    if (!event) throw new NotFoundException('Event not found');
    event.participantsCount++;
    return event;
  }

  // Analytics
  getAnalytics() {
    return {
      totalReferrals: this.referrals.size,
      totalCodes: this.referralCodes.size,
      totalEvents: this.events.size,
      activeEvents: Array.from(this.events.values()).filter(e => e.status === 'active').length,
    };
  }
}

File: lib/features/growth/services/growth_service.dart
import 'dart:convert';
import 'dart:math';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/referral_models.dart';
import '../models/gamification_models.dart';
import '../models/seasonal_event.dart';
import '../models/push_notification.dart';
import '../models/ab_testing.dart';
import '../models/premium_conversion.dart';

class GrowthService {
  static const String _referralKey = 'referral_data';
  static const String _progressKey = 'user_progress';
  static const String _eventsKey = 'seasonal_events';
  static const String _notificationsKey = 'push_notifications';
  static const String _abTestsKey = 'ab_tests';
  static const String _premiumKey = 'premium_data';

  // Referral System
  Future<ReferralCode> generateReferralCode(String userId) async {
    final code =
        'LEXI${userId.substring(0, min(4, userId.length)).toUpperCase()}${Random().nextInt(9999).toString().padLeft(4, '0')}';

    final referralCode = ReferralCode(
      id: 'ref_${DateTime.now().millisecondsSinceEpoch}',
      userId: userId,
      code: code,
      maxUses: 10,
      createdAt: DateTime.now(),
      expiresAt: DateTime.now().add(const Duration(days: 30)),
    );

    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      '${_referralKey}_code_$userId',
      jsonEncode(referralCode.toJson()),
    );

    return referralCode;
  }

  Future<ReferralStats> getReferralStats(String userId) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final statsJson = prefs.getString('${_referralKey}_stats_$userId');
      if (statsJson != null) {
        return ReferralStats.fromJson(jsonDecode(statsJson));
      }
    } catch (e) {
      // Fall through to default
    }
    return ReferralStats.empty();
  }

  Future<Referral?> applyReferralCode(
    String code,
    String inviteeId,
    String inviteeName,
  ) async {
    final prefs = await SharedPreferences.getInstance();

    // Find the referral code
    final keys = prefs.getKeys().where(
      (k) => k.startsWith('${_referralKey}_code_'),
    );
    for (final key in keys) {
      final codeJson = prefs.getString(key);
      if (codeJson != null) {
        final referralCode = ReferralCode.fromJson(jsonDecode(codeJson));
        if (referralCode.code == code && referralCode.canUse) {
          // Create referral
          final referral = Referral(
            id: 'ref_${DateTime.now().millisecondsSinceEpoch}',
            referrerId: referralCode.userId,
            referrerName: 'User',
            inviteeId: inviteeId,
            inviteeName: inviteeName,
            referralCode: code,
            status: ReferralStatus.accepted,
            createdAt: DateTime.now(),
            acceptedAt: DateTime.now(),
            reward: const ReferralReward(gems: 500, premiumDays: 7),
          );

          // Update code usage
          final updatedCode = referralCode.copyWith(
            currentUses: referralCode.currentUses + 1,
          );
          await prefs.setString(key, jsonEncode(updatedCode.toJson()));

          return referral;
        }
      }
    }
    return null;
  }

  // Gamification
  Future<UserProgress> getUserProgress(String userId) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final progressJson = prefs.getString('${_progressKey}_$userId');
      if (progressJson != null) {
        return UserProgress.fromJson(jsonDecode(progressJson));
      }
    } catch (e) {
      // Fall through to default
    }
    return UserProgress.empty();
  }

  Future<void> saveUserProgress(String userId, UserProgress progress) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      '${_progressKey}_$userId',
      jsonEncode(progress.toJson()),
    );
  }

  Future<void> addXp(String userId, int xp) async {
    final progress = await getUserProgress(userId);
    final newTotalXp = progress.totalXp + xp;
    final newLevel = _calculateLevel(newTotalXp);

    final updatedProgress = progress.copyWith(
      totalXp: newTotalXp,
      currentLevel: newLevel,
    );

    await saveUserProgress(userId, updatedProgress);
  }

  int _calculateLevel(int xp) {
    int level = 1;
    int requiredXp = 100;

    while (xp >= requiredXp) {
      level++;
      requiredXp += (level * 50);
    }

    return level;
  }

  List<UserLevel> getAllLevels() {
    return [
      const UserLevel(
        level: 1,
        title: 'Beginner',
        minXp: 0,
        maxXp: 100,
        reward: LevelReward(gems: 0),
      ),
      const UserLevel(
        level: 5,
        title: 'Explorer',
        minXp: 500,
        maxXp: 1000,
        reward: LevelReward(gems: 100, frameId: 'explorer_frame'),
      ),
      const UserLevel(
        level: 10,
        title: 'Adventurer',
        minXp: 2000,
        maxXp: 3000,
        reward: LevelReward(gems: 200, frameId: 'adventurer_frame'),
      ),
      const UserLevel(
        level: 15,
        title: 'Scholar',
        minXp: 5000,
        maxXp: 7500,
        reward: LevelReward(gems: 300, badgeId: 'scholar_badge'),
      ),
      const UserLevel(
        level: 20,
        title: 'Expert',
        minXp: 10000,
        maxXp: 15000,
        reward: LevelReward(gems: 500, frameId: 'expert_frame'),
      ),
      const UserLevel(
        level: 25,
        title: 'German Student',
        minXp: 20000,
        maxXp: 30000,
        reward: LevelReward(gems: 750, badgeId: 'german_student_badge'),
      ),
      const UserLevel(
        level: 30,
        title: 'Linguist',
        minXp: 40000,
        maxXp: 60000,
        reward: LevelReward(gems: 1000, frameId: 'linguist_frame'),
      ),
      const UserLevel(
        level: 40,
        title: 'Master',
        minXp: 100000,
        maxXp: 150000,
        reward: LevelReward(gems: 2000, badgeId: 'master_badge'),
      ),
      const UserLevel(
        level: 50,
        title: 'Language Master',
        minXp: 250000,
        maxXp: 300000,
        reward: LevelReward(gems: 5000, frameId: 'master_frame'),
      ),
    ];
  }

  List<AchievementBadge> getAllBadges() {
    return [
      const AchievementBadge(
        id: 'first_lesson',
        name: 'First Steps',
        description: 'Complete your first lesson',
        icon: '🎯',
        rarity: BadgeRarity.common,
      ),
      const AchievementBadge(
        id: 'streak_7',
        name: 'Week Warrior',
        description: '7-day streak',
        icon: '🔥',
        rarity: BadgeRarity.uncommon,
      ),
      const AchievementBadge(
        id: 'streak_30',
        name: 'Monthly Master',
        description: '30-day streak',
        icon: '💪',
        rarity: BadgeRarity.rare,
      ),
      const AchievementBadge(
        id: 'words_100',
        name: 'Word Collector',
        description: 'Learn 100 words',
        icon: '📚',
        rarity: BadgeRarity.uncommon,
      ),
      const AchievementBadge(
        id: 'words_1000',
        name: 'Vocabulary Master',
        description: 'Learn 1000 words',
        icon: '🎓',
        rarity: BadgeRarity.epic,
      ),
      const AchievementBadge(
        id: 'speaking_100',
        name: 'Speaking Star',
        description: 'Practice speaking 100 minutes',
        icon: '🎤',
        rarity: BadgeRarity.rare,
      ),
      const AchievementBadge(
        id: 'goethe_a1',
        name: 'Goethe A1',
        description: 'Pass Goethe A1 exam',
        icon: '📝',
        rarity: BadgeRarity.uncommon,
      ),
      const AchievementBadge(
        id: 'goethe_b1',
        name: 'Goethe B1',
        description: 'Pass Goethe B1 exam',
        icon: '📝',
        rarity: BadgeRarity.rare,
      ),
      const AchievementBadge(
        id: 'referral_5',
        name: 'Social Butterfly',
        description: 'Invite 5 friends',
        icon: '🦋',
        rarity: BadgeRarity.epic,
      ),
      const AchievementBadge(
        id: 'summer_master',
        name: 'Summer Master',
        description: 'Complete Summer Challenge',
        icon: '☀️',
        rarity: BadgeRarity.legendary,
      ),
    ];
  }

  // Seasonal Events
  Future<List<SeasonalEvent>> getSeasonalEvents() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final eventsJson = prefs.getString(_eventsKey);
      if (eventsJson != null) {
        final List<dynamic> list = jsonDecode(eventsJson);
        return list.map((j) => SeasonalEvent.fromJson(j)).toList();
      }
    } catch (e) {
      debugPrint('Failed to load cached seasonal events: $e');
    }
    return [];
  }

  Future<void> saveSeasonalEvents(List<SeasonalEvent> events) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      _eventsKey,
      jsonEncode(events.map((e) => e.toJson()).toList()),
    );
  }

  // Push Notifications
  Future<List<PushNotification>> getNotifications(String userId) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final notificationsJson = prefs.getString('${_notificationsKey}_$userId');
      if (notificationsJson != null) {
        final List<dynamic> list = jsonDecode(notificationsJson);
        return list.map((j) => PushNotification.fromJson(j)).toList();
      }
    } catch (e) {
      debugPrint('Failed to load cached notifications: $e');
    }
    return [];
  }

  List<PushNotification> _getDemoNotifications() {
    return [
      PushNotification(
        id: 'notif_1',
        title: 'Streak Alert 🔥',
        body: 'Don\'t lose your 7-day streak! Practice for just 10 minutes.',
        type: NotificationType.streak,
        priority: NotificationPriority.high,
        scheduledAt: DateTime.now().subtract(const Duration(hours: 2)),
        sentAt: DateTime.now().subtract(const Duration(hours: 2)),
      ),
      PushNotification(
        id: 'notif_2',
        title: 'New Achievement! 🏆',
        body: 'You unlocked "Week Warrior" badge!',
        type: NotificationType.achievement,
        priority: NotificationPriority.medium,
        scheduledAt: DateTime.now().subtract(const Duration(days: 1)),
        sentAt: DateTime.now().subtract(const Duration(days: 1)),
        isRead: true,
      ),
      PushNotification(
        id: 'notif_3',
        title: 'Challenge Available 🎯',
        body: 'Join the German Summer Challenge and win exclusive rewards!',
        type: NotificationType.challenge,
        priority: NotificationPriority.medium,
        scheduledAt: DateTime.now().subtract(const Duration(days: 2)),
        sentAt: DateTime.now().subtract(const Duration(days: 2)),
        isRead: true,
      ),
    ];
  }

  List<NotificationTemplate> getSmartTemplates() {
    return [
      const NotificationTemplate(
        id: 'template_streak_risk',
        title: '{name}, your streak is waiting 🔥',
        body:
            'Only {minutes} minutes today to continue your {streak}-day streak!',
        type: NotificationType.streak,
        priority: NotificationPriority.high,
        variables: ['name', 'minutes', 'streak'],
        condition: 'streak > 0 && daysSinceActive > 0',
      ),
      const NotificationTemplate(
        id: 'template_comeback',
        title: 'We miss you, {name}! 💙',
        body:
            'Your German learning is waiting. Just {minutes} minutes a day makes a difference.',
        type: NotificationType.reminder,
        priority: NotificationPriority.medium,
        variables: ['name', 'minutes'],
        condition: 'daysSinceActive > 3',
      ),
      const NotificationTemplate(
        id: 'template_achievement',
        title: 'Achievement Unlocked! 🏆',
        body: 'You earned "{badge_name}"! Keep up the great work.',
        type: NotificationType.achievement,
        priority: NotificationPriority.medium,
        variables: ['badge_name'],
        condition: 'newAchievement',
      ),
      const NotificationTemplate(
        id: 'template_premium_trial',
        title: 'Try Premium Free! ✨',
        body: '{name}, unlock all features with a {days}-day free trial.',
        type: NotificationType.premium,
        priority: NotificationPriority.low,
        variables: ['name', 'days'],
        condition: '!isPremium && trialAvailable',
      ),
    ];
  }

  // A/B Tests
  Future<List<ABTest>> getABTests() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final testsJson = prefs.getString(_abTestsKey);
      if (testsJson != null) {
        final List<dynamic> list = jsonDecode(testsJson);
        return list.map((j) => ABTest.fromJson(j)).toList();
      }
    } catch (e) {
      debugPrint('Failed to load cached A/B tests: $e');
    }
    return [];
  }

  ABTestVariant getVariant(String testId, String userId) {
    final hash = (testId + userId).hashCode;
    return hash.isEven ? ABTestVariant.a : ABTestVariant.b;
  }

  // Premium
  Future<SmartPaywall> getSmartPaywall(String userId) async {
    final progress = await getUserProgress(userId);
    final context = {
      'lessonsCompleted': progress.stats['lessonsCompleted'] ?? 0,
      'wordsLearned': progress.stats['wordsLearned'] ?? 0,
      'streak': progress.streak,
      'totalMinutes': progress.stats['totalMinutes'] ?? 0,
    };

    return SmartPaywall.personalized(context);
  }

  Future<FreeTrial> startFreeTrial(String userId) async {
    final trial = FreeTrial(
      id: 'trial_${DateTime.now().millisecondsSinceEpoch}',
      durationDays: 7,
      startDate: DateTime.now(),
      endDate: DateTime.now().add(const Duration(days: 7)),
      planAfterTrial: PremiumPlan.yearly,
    );

    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      '${_premiumKey}_trial_$userId',
      jsonEncode(trial.toJson()),
    );

    return trial;
  }
}
